Ch1 revision (#143)
UBC-DSCI/introduction-to-datascience: index.Rmd

  readr::write_csv("data/can_lang.csv")
```

# R and the tidyverse
## Overview

This textbook aims to be an approachable introduction to the world of data science. In this book, we define *data science* as the study and development of reproducible, auditable processes to obtain value (i.e., insight) from data. At a high level, in this book, readers will learn how to 

(1) use reproducible tools to do data analysis and 
(2) identify and solve common problems in data science. 

> The book is structured, so learners spend the first four chapters learning how to use the R programming language to load, wrangle, clean, and visualize data while answering descriptive and exploratory data analysis questions. The next six chapters illustrate how to solve common problems in data science, which help answer predictive and inferential data analysis questions. In the final chapter, we discuss how to set up your own computer with tools like R and Jupyter. For readers who need help setting up their system on their computer, we recommend starting with the [Setting up your own computer] chapter before reading chapter 2. Figure \@ref(fig:img-chapter-overview) summarizes a data science pipeline and what you will learn in each chapter of this book.

Reviewer: We will 
```suggestion
The book is structured, so learners spend the first four chapters learning how to use the R programming language to load, wrangle, clean, and visualize data while answering descriptive and exploratory data analysis questions. The next six chapters illustrate how to solve common problems in data science, which help answer predictive and inferential data analysis questions. In the final chapters (\@ref(getting-started-with-jupyter) - \@ref(setting-up-your-own-computer)), we discuss how to use Jupyter to edit and write R code, how to use version control for collaboration, and how to install and configure the software needed for data science on your own computer. Readers may want to refer to these chapters earlier as they start exploring the examples and exercises in this book. Figure \@ref(fig:img-chapter-overview) summarizes a data science pipeline and what you will learn in each chapter of this book.
```

```{r img-chapter-overview, echo = FALSE, message = FALSE, warning = FALSE, fig.cap = "Where are we going?", out.width="1050", fig.retina = 2}
knitr::include_graphics("img/chapter_overview.jpeg")
```

> The four common data science problems that are useful for answering predictive and inferential data analysis questions covered in this book are:

Reviewer: Problems are not useful for answering questions. Let's reword to:
```suggestion
The four common data science methods that are useful for answering predictive and inferential data analysis questions covered in this book are:
```

(2) identify and solve common problems in data science. 

The book is structured, so learners spend the first four chapters learning how to use the R programming language to load, wrangle, clean, and visualize data while answering descriptive and exploratory data analysis questions. The next six chapters illustrate how to solve common problems in data science, which help answer predictive and inferential data analysis questions. In the final chapter, we discuss how to set up your own computer with tools like R and Jupyter. For readers who need help setting up their system on their computer, we recommend starting with the [Setting up your own computer] chapter before reading chapter 2. Figure \@ref(fig:img-chapter-overview) summarizes a data science pipeline and what you will learn in each chapter of this book.

```{r img-chapter-overview, echo = FALSE, message = FALSE, warning = FALSE, fig.cap = "Where are we going?", out.width="1050", fig.retina = 2}
knitr::include_graphics("img/chapter_overview.jpeg")
```

The four common data science problems that are useful for answering predictive and inferential data analysis questions covered in this book are:

> 1. **Classification: predicting a class/category for a new observation.** This could answer a question such as whether a tumour is cancerous or benign given some tumour cell measurements?

Reviewer: Edited in 850c001

Author: This change looks good! Thanks Tiffany!

2. **Regression: predicting a quantitative value for a new observation.** This could answer a question such as what will be the 10 km race time for 20-year-old woman, with a BMI of 25 who trains 4 times a week?

3. **Clustering: finding previously unknown/unlabelled subgroups in a dataset**. This could answer a question such as what products commonly bought together on Amazon?

4. **Estimation: make a good guess of an average or a proportion for the wider population from a representative sample (group of people or units), and quantify how good that guess is.** This could answer a question such as what is the proportion of undergraduate students that own an iPhone?

> We map each of these problems to the type of data analysis question and discuss what kinds of data are needed to answer such questions [@leek2015question; @peng2015art]. More advanced (e.g., causal or mechanistic) data analysis questions are beyond the scope of this text.

Reviewer: ```suggestion
We map each of these methods to the broader type of data analysis question they are useful for answering, as well as discuss what kinds of data are needed to answer such questions [@leek2015question; @peng2015art]. More advanced (*e.g.,* causal or mechanistic) data analysis questions are beyond the scope of this text.
```

| Causal | A question that asks about whether changing one factor will lead to a change in another factor, on average, in the wider population. | Does wealth lead to voting for a certain political party in Canadian elections? |
| Mechanistic | A question that asks about the underlying mechanism of the observed patterns, trends, or relationships (i.e., how does it happen?) | How does wealth lead to voting for a certain political party in Canadian elections? |

Source: [What is the question?](https://science.sciencemag.org/content/347/6228/1314) by Jeffery T. Leek, Roger D. Peng & [The Art of Data Science](https://leanpub.com/artofdatascience) by Roger Peng & Elizabeth Matsui

## Chapter Learning Objectives

By the end of the chapter, students will be able to:

- load the `tidyverse` package into R
- create new variables and objects in R using the assignment symbol
- use the help and documentation tools in R
- match the names of the following functions from the `tidyverse` package to their documentation descriptions: 
    - `read_csv` 
    - `select`
    - `filter`
    - `ggplot`
    - `aes`

## Loading a spreadsheet-like data set

> Often, the first step in data analysis is to load a data set into R. When we bring spreadsheet-like (think Microsoft Excel tables) data, generally shaped like a rectangle, into R, it is represented as a *data frame* object. In figure \@ref(fig:img-spreadsheet-vs-dataframe), we can see an R data frame is very similar to a spreadsheet where the rows are the collected observations, and the columns are the variables. 

Reviewer: I think we can be more confident that this is almost always the first step.
```suggestion
Almost always, the first step in data analysis is to load a data set into R. When we bring rectangular-shaped, spreadsheet-like data (think Microsoft Excel tables) into R, it is represented as a *data frame* object. In figure \@ref(fig:img-spreadsheet-vs-dataframe), we can see an R data frame is very similar to a spreadsheet where the rows are the collected observations, and the columns are the variables. 
```

  readr::write_csv("data/can_lang.csv")
```

# R and the tidyverse
## Overview

> This textbook aims to be an approachable introduction to the world of data science. In this book, we define *data science* as the study and development of reproducible, auditable processes to obtain value (i.e., insight) from data. At a high level, in this book, readers will learn how to 

Reviewer: This is a really long line. We should use line breaks to limit the characters per line to 80 characters (RStudio has a nice vertical line that tells you where this is). In addition to limiting to 80 characters, I have just started learning about semantic line breaks (see https://sembr.org/) in which you add a line break after each substantial unit of thought. I think if we do this in our writing, our revisions will become easier to manage.

(1) use reproducible tools to do data analysis and 
(2) identify and solve common problems in data science. 

> The book is structured, so learners spend the first four chapters learning how to use the R programming language to load, wrangle, clean, and visualize data while answering descriptive and exploratory data analysis questions. The next six chapters illustrate how to solve common problems in data science, which help answer predictive and inferential data analysis questions. In the final chapter, we discuss how to set up your own computer with tools like R and Jupyter. For readers who need help setting up their system on their computer, we recommend starting with the [Setting up your own computer] chapter before reading chapter 2. Figure \@ref(fig:img-chapter-overview) summarizes a data science pipeline and what you will learn in each chapter of this book.

Reviewer: We should formally cite the R programming language. The citation is pasted below and it should be put in `references.bib`. Once that is done we can reference it inline via `@Rlanguage`:

```
  @Manual{Rlanguage,
    title = {R: A Language and Environment for Statistical Computing},
    author = {{R Core Team}},
    organization = {R Foundation for Statistical Computing},
    address = {Vienna, Austria},
    year = {2021},
    url = {https://www.R-project.org/},
  }
```

| Causal | A question that asks about whether changing one factor will lead to a change in another factor, on average, in the wider population. | Does wealth lead to voting for a certain political party in Canadian elections? |
| Mechanistic | A question that asks about the underlying mechanism of the observed patterns, trends, or relationships (i.e., how does it happen?) | How does wealth lead to voting for a certain political party in Canadian elections? |

Source: [What is the question?](https://science.sciencemag.org/content/347/6228/1314) by Jeffery T. Leek, Roger D. Peng & [The Art of Data Science](https://leanpub.com/artofdatascience) by Roger Peng & Elizabeth Matsui

## Chapter Learning Objectives

By the end of the chapter, students will be able to:

- load the `tidyverse` package into R
- create new variables and objects in R using the assignment symbol
- use the help and documentation tools in R
- match the names of the following functions from the `tidyverse` package to their documentation descriptions: 
    - `read_csv` 
    - `select`
    - `filter`
    - `ggplot`

>     - `aes`

Reviewer: Should we add `slice` here?

## Loading a spreadsheet-like data set

Often, the first step in data analysis is to load a data set into R. When we bring spreadsheet-like (think Microsoft Excel tables) data, generally shaped like a rectangle, into R, it is represented as a *data frame* object. In figure \@ref(fig:img-spreadsheet-vs-dataframe), we can see an R data frame is very similar to a spreadsheet where the rows are the collected observations, and the columns are the variables. 

```{r img-spreadsheet-vs-dataframe, echo = FALSE, message = FALSE, warning = FALSE, fig.cap = "A spreadsheet versus a data frame in R", out.width="850", fig.retina = 2}
knitr::include_graphics("img/spreadsheet_vs_dataframe.PNG")
```

The first kind of data file that we will learn how to load into R (as a data frame) is the 
spreadsheet-like *comma-separated values* format (`.csv` for short).
These files have names ending in `.csv`, and can be opened and saved using common spreadsheet programs like Microsoft Excel and Google Sheets. 
For example, a `.csv` file named `can_lang.csv` [is included with the code for this book](https://github.com/UBC-DSCI/introduction-to-datascience/blob/master/data/can_lang.csv). 

> This file--- originally from [{canlang} R data package](https://ttimbers.github.io/canlang/)---has language data collected in the 2016 Canadian census [@cancensus2016]. 

Reviewer: We should formally cite all R packages. The citation for this one is here: https://ttimbers.github.io/canlang/authors.html

Author: Good point! Will add!

(2) identify and solve common problems in data science. 

The book is structured, so learners spend the first four chapters learning how 
to use the R programming language [@Rlanguage] to load, wrangle, clean, and visualize data
while answering descriptive and exploratory data analysis questions. The next
six chapters illustrate how to answer
predictive, exploratory and inferential data analysis questions with common methods
in data science, including classification, regression, clustering and estimation.
In the final chapters (\@ref(getting-started-with-jupyter) - \@ref(setting-up-your-own-computer)), 
we discuss how to use Jupyter to edit and write R code, how to use version control for collaboration, and how to install and configure the software needed for data science on your own computer. Readers 
may want to refer to these chapters earlier as they start exploring the examples 
and exercises in this book. Figure \@ref(fig:img-chapter-overview) summarizes a data science pipeline and what you will learn in each chapter of this book.

```{r img-chapter-overview, echo = FALSE, message = FALSE, warning = FALSE, fig.cap = "Where are we going?", out.width="1050", fig.retina = 2}

> knitr::include_graphics("img/chapter_overview.jpeg")

Author: Made a new issue #178 so that we can merge this branch

| Exploratory | A question asks if there are patterns, trends, or relationships within a single data set. Often used to propose hypotheses for future study. | Does political party voting change with indicators of wealth in a set of data collected on 2,000 people living in Canada? |
| Inferential | A question that looks for patterns, trends, or relationships in a single data set **and** also asks for quantification of how applicable these findings are to the wider population. | Does political party voting change with indicators of wealth for all people living in Canada? |
| Predictive | A question that asks about predicting measurements or labels for individuals (people or things). The focus is on what things predict some outcome, but not what causes the outcome. | What political party will someone vote for in the next Canadian election? |
| Causal | A question that asks about whether changing one factor will lead to a change in another factor, on average, in the wider population. | Does wealth lead to voting for a certain political party in Canadian elections? |

> | Mechanistic | A question that asks about the underlying mechanism of the observed patterns, trends, or relationships (i.e., how does it happen?) | How does wealth lead to voting for a certain political party in Canadian elections? |

Author: Making new issue #179

## Chapter Learning Objectives

By the end of the chapter, students will be able to:

- load the `tidyverse` package into R
- create new variables and objects in R using the assignment symbol
- use the help and documentation tools in R
- match the names of the following functions from the `tidyverse` package to their documentation descriptions: 
    - `read_csv` 
    - `select`
    - `filter`
    - `arrange`
    - `slice`
    - `ggplot`
    - `aes`

## Ask a question 
Where to begin? As outlined in Figure \@ref(fig:img-chapter-overview), we will start
with a question we want to answer. Generally, we need some context to frame our question. 

Before the arrival of colonists, Indigenous peoples lived in Canada with their
own cultures and many different languages. Many Aboriginal languages are unique
to Canada and not spoken anywhere else in the world [@statcan2018mothertongue].
Sadly, colonization has led to the loss of many Aboriginal languages. For instance,
generations of children were not allowed to speak their mother tongue
(the first language an individual learns in childhood) in Canadian residential 

> schools. Colonizers also renamed places they had "discovered" [@wilson2018]. Acts such as these have significantly harmed the continuity of Aboriginal languages in Canada, and some languages are considered "endangered" as few people report speaking them. 

Author: made new issue about this #177

```{r img-spreadsheet-vs-dataframe, echo = FALSE, message = FALSE, warning = FALSE, fig.cap = "A spreadsheet versus a data frame in R", out.width="850", fig.retina = 2}
knitr::include_graphics("img/spreadsheet_vs_dataframe.PNG")
```

The first kind of data file that we will learn how to load into R (as a data frame) is the 
spreadsheet-like *comma-separated values* format (`.csv` for short).
These files have names ending in `.csv`, and can be opened and saved using common spreadsheet programs like Microsoft Excel and Google Sheets. 
For example, the `.csv` file named `can_lang.csv` [is included with the code for this book](https://github.com/UBC-DSCI/introduction-to-datascience/blob/master/data/can_lang.csv). 
If we were to open this data in a plain text editor, we would see each row on its own line, and each entry in the table separated by a comma:

```{bash, echo=FALSE, comment=NA}
head -n 10 data/can_lang.csv
```

> To load this data into R, so that we can do things with it (*e.g,* perform analyses or create data visualizations), we will need to use a *function.* 

Author: Not sure what you mean by this? @ttimbers

- uses a comma (`,`) to separate the columns, and
- does not have row names.

Below you'll see the code used to load the data into R using the `read_csv` 
function. However, given that `read_csv` is not included in the base 
installation of R, we will have to load it from somewhere else before we can use
it. The place we will load it from is an R package. An R package is a collection
of functions that once loaded, can be used in addition to the built-in R package
functions. The `read_csv` function, in particular, can be made accessible 

> through loading the `tidyverse` package (more on this later), using the 

Reviewer: We should formally cite the tidyverse R package. The citation is pasted below: 

```
  @Article{,
    title = {Welcome to the {tidyverse}},
    author = {Hadley Wickham and Mara Averick and Jennifer Bryan and Winston Chang and Lucy D'Agostino McGowan and Romain François and Garrett Grolemund and Alex Hayes and Lionel Henry and Jim Hester and Max Kuhn and Thomas Lin Pedersen and Evan Miller and Stephan Milton Bache and Kirill Müller and Jeroen Ooms and David Robinson and Dana Paige Seidel and Vitalie Spinu and Kohske Takahashi and Davis Vaughan and Claus Wilke and Kara Woo and Hiroaki Yutani},
    year = {2019},
    journal = {Journal of Open Source Software},
    volume = {4},
    number = {43},
    pages = {1686},
    doi = {10.21105/joss.01686},
  }
```

> Note: you can find most CRAN R package citations via `citation("package_name")`

`library` function.

After loading the `tidyverse` package, we can call the `read_csv` function and 
pass it a single argument: the name of the file, `"can_lang.csv"`. We have to put quotes around file names and other letters and words that we use in our code to distinguish it from the special words that make up the R programming language.  The file's name is the only argument we need to provide because our file satisfies everything else the `read_csv` function expects in the default use-case (which we just discussed). Later in the course, we'll learn more about dealing with more complicated files where the default arguments are not appropriate, for example, files that use spaces or tabs to separate the columns instead of commas.

```{r load_state_property_data, warning=FALSE, message=FALSE}
library(tidyverse)
read_csv("data/can_lang.csv")
```

> **In case you want to know more (optional):** 
> We use the `read_csv` function from the `tidyverse` instead of the base R function `read.csv` because it's faster, and it creates a nicer variant of the base R data frame called a *tibble*. 
> This has several benefits that we'll discuss in further detail later in the course.

> ## Giving a data frame a name

Reviewer: Could we call this, "Naming things"? This topic is more general than just data frames. Then we could also add a couple sentences on what are good names? E.g., they should be syntactic (you don't need to use this word, but summarize what qualifies as this, as specified here: https://principles.tidyverse.org/names-attribute.html#universal-names) and recommend that we follow the tidyverse naming convention in this book (snake_case).

> When we loaded the language data collected in the 2016 Canadian census in R above using `read_csv`, we did not give this data frame a name. Therefore the data was just printed on the screen, and we cannot do anything else with it. That isn't very useful; we would like to give a name to the data frame that `read_csv` outputs to use later for analysis and visualization. 

Reviewer: ```suggestion
When we loaded the language data collected in the 2016 Canadian census in R above using `read_csv`, 
we did not give this data frame a name. Therefore the data was just printed on the screen, 
and we cannot do anything else with it. That isn't very useful. What would be more useful
would be to give a name to the data frame that `read_csv` outputs, 
so that we could refer to it later for analysis and visualization. 
```

> There are two possible ways to assign a name to something in R ---using either the assignment symbol (`<-`) or the equals symbol (`=`). From a style perspective, the assignment symbol is preferred and is what we will use in this course. When we name something in R utilizing the assignment symbol, `<-`, we do not need to surround it with quotes like the file name because we are formally telling R about this word and giving it a value. Only characters and words that act as values need to be surrounded by quotes.

Reviewer: ```suggestion
There are two possible ways to assign a name to something in R ---
using either the assignment symbol (`<-`) or the equals symbol (`=`). 
From a style perspective, the assignment symbol is preferred 
and is what we will use in this course. 
When we name something in R using the assignment symbol, `<-`, 
we do not need to surround the name we are giving (on `<-`'s left-hand side) 
with quotes like the file name 
because we are formally telling R about this new word and giving it a value. 
Only characters and words that act as values (on `<-`'s right-hand side) 
need to be surrounded by quotes.
```

> Let's now use the assignment symbol to give the name `can_lang` to the language data collected in the 2016 Canadian census data frame that we get from `read_csv`.

Reviewer: ```suggestion
Let's now use the assignment symbol to give the name `can_lang` 
to the 2016 Canadian census language data frame that we get from `read_csv`. 
```

```{r load_data_with_name, message=FALSE}
can_lang <- read_csv("data/can_lang.csv")
```

> Wait a minute! Nothing happened this time! Or at least it looks like that. But actually, something did happen: the data was read in and now has the name `can_lang` associated with it. And we can use that name to access the data frame and do things with it. We can type the name of the data frame to print the first few rows on the screen.

Reviewer: ```suggestion
Wait a minute, nothing happened this time! Or at least it looks like that? 
But actually, something did happen! The data was read in 
and now has the name `can_lang` associated with it. 
And we can use that name to access the data frame and do things with it. 
We can type the name of the data frame to print the first few rows 
on the screen.
```

```{r print}
can_lang
```

> ## Creating subsets of data frames with `select` & `filter`

Reviewer: Can we do something interesting here with `select` and `filter`? Perhaps something that will lead us to create our final visualization in this chapter?

Perhaps we can state our question here, which is "what ten Aboriginal languages were most often reported in 2016 as mother tongues in Canada and how many people speak each of them?" and then try to answer them first by creating a table with `filter` & `select`? And then we can bring up the idea that although tables are great, visualizations are often even more insightful. And then we make the visualization?

So if we did that, then the story would go first `filter` to get the Aboriginal languages, and then `select` just `language` and `mother_tongue`? Then we do `slice` to get the top 10? I suggest `slice` instead of `slice_max` as we use `slice` a lot in later worksheets for the course, and it is a bit more general. We have to use `reorder` for the plot, and then we might get questions, why doesn't `slice_max` change the bar order? Which is a good question, but I don't think we want them to be asking that yet.

>Note: another thing that I think we do want to add in here is a richer description introducing the question we are asking. Perhaps something along the lines of Aboriginals were in Canada before colonization, and there were many groups with X different languages. Sadly, colonization has led to the loss of many of these. In the 2016 census, Y Aboriginal languages were recorded as still being spoken. To learn more see this resource (this Canadian/National Geographic reference looks really good: https://www.canadiangeographic.ca/article/mapping-indigenous-languages-canada and maybe another?).

Reviewer: @leem44 - Now wondering if we want to start off the chapter with our question? Specifically, just before we teach them to load data. Then we have a question we want to answer, some data in a file that can answer it. Now we just need to use R to load the data and do some wrangling and viz to answer it!

Author: I really like this idea! I will make changes

### Using `select` to extract columns

Now let's use `select` to extract the `language` and `mother_tongue` columns from this data frame. To extract tese columns, we need to provide the `select` function with three arguments. The first argument is the 
name of the data frame object, which in this example is `aboriginal_lang`. The second and third arguments are the column names that we want to select, here `language` and `mother_tongue`. After passing these three arguments,
the  `select` function returns two columns (the `language` and `mother_tongue` columns that we asked for) as a data frame.

```{r}
selected_lang <- select(aboriginal_lang, language, mother_tongue)
selected_lang
```
> Note: we didn't actually *need* to `select` the columns in our data frame before moving on to the next step! However, `select` allows us to display only the columns in the data frame we want and can make it easier for us to view the data especially if we have a large data frame with lots of columns. 

### Using `arrange` to order and `slice` to select rows by index number

> The  `arrange` function allows us to order the rows of a data frame by the values of a selected column. We need to pass the data frame, and the variable to order by in the first two arguments of this function. Since we want to choose the ten Aboriginal languages most often reported as a mother tongue language, we will use the `arrange` function to order the rows in our `selected_lang` data frame by the `mother_tongue` column. We want to order the rows in descending order (from largest to smallest) so we include `desc` before the column name `mother_tongue` to sort in descending order. 

Author: Yes that is a really good idea! I will make that change!

```{r}
arranged_lang <- arrange(selected_lang, by = desc(mother_tongue))
arranged_lang
```

Next we will use the `slice` function, which selects rows according to their row number. Since we want to choose ten languages, we will indicate we want the rows 1 to 10 using the argument `1:10`.
```{r}
ten_lang <- slice(arranged_lang, 1:10)
ten_lang
```

> We have now answered our initial question by generating this frequency table! Are we done? Technically, this table answers our question. However, we can go one step further and create a visualization to answer our question as well. Visualizations are a great tool for summarizing information to help you effectively communicate with your audience. While tables are great for displaying information, a visualization can sometimes provide more insight. For example, if we were really interested in *comparing* the values between the different languages (e.g. how many more people reported "Cree, n.o.s." versus "Inuktitut" versus "Ojibway" etc.), then a visualization is a more effective way to communicate your results than a frequency table. 

Author: Thanks!

### Using `ggplot` to create a bar plot

In our data set, we can see that each variable--- higher-level language category (`category`), language (`language`), mother tongue (`mother_tongue`), language spoken most at home (`most_at_home`), etc. ---are all in separate columns. In addition, there is a single row (or observation) for each language. The data are, therefore, in what we call a *tidy data* format. Tidy data is a fundamental concept and will be a significant focus in the remainder of this course: many of the functions from `tidyverse` require tidy data, including the `ggplot` function that we will use shortly for our visualization. We will formally introduce this concept in chapter 3.

We will make a bar plot using the `most_at_home` and `language` columns from our `ten_lang` data frame. To create a bar plot of these two variables using the `ggplot` function, we would the following: 

> ```{r img-ggplot, echo = FALSE, message = FALSE, warning = FALSE, fig.cap = "Creating a bar plot with the ggplot function", out.width="1100", fig.retina = 2}

Author: I agree! I wanted to do these for other functions, but it takes a bit of time to put together. I will make an issue for this and hopefully, we can add these after the global tasks are done!

knitr::include_graphics("img/ggplot_function.jpeg")
```

> Below we create a bar plot using the `ggplot` function following the instructions described in figure \@ref(fig:img-ggplot):

Reviewer: Also, I think the bar charts take up more space than they need... They could be 75% of their height at least.

### Putting it all together

In the code chunk below, we put everything from this chapter together. We have added a few more layers to make the data visualization even more effective. Specifically, we changed the colour of the bars and changed the background from grey to white to improve the contrast. Note that we actually skipped the `select` step that we did above just to show you that we didn't actually need to `select` the columns to create the visualization! 

> Notice that we provided *comments* beside some of the code below using the hashtag symbol `#`. You can use comments to explain lines of code for others or yourself in the future! If R sees a `#` sign, it will ignore all the code that comes after it on that line. It's good practice to get in the habit of commenting your code to improve your code's readability.

```{r nachos-to-cheesecake, fig.width=7.75, fig.height=4, warning=FALSE, message=FALSE, fig.cap = "Putting it all together: Bar plot of the ten Aboriginal languages most often reported by Canadians as their mother tongue"}

> library(tidyverse)

Reviewer: There is too much whitespace between lines in the code here. I would do it like this instead:

```
library(tidyverse)

can_lang <- read_csv("data/can_lang.csv")

aboriginal_lang <- filter(can_lang, category == "Aboriginal languages")
arranged_lang <- arrange(aboriginal_lang, by = desc(mother_tongue))
ten_lang <- slice(arranged_lang, 1:10)

ggplot(ten_lang, aes(
  x = reorder(language, mother_tongue),
  y = mother_tongue
)) +
  geom_bar(stat = "identity", fill = "steelblue") +
  xlab("Language") +
  ylab("Mother tongue (Number of Canadians residents)") +
  coord_flip() +
  theme_bw() # use a theme to have a white background
```

To answer our data analysis question, we will begin by using the `slice_max` function, which selects rows with the highest values of a variable. Here we want to choose the ten Aboriginal languages most often reported as a mother tongue language. In the function, we need to pass the data frame, the number of rows to select, and the variable to order by. 

```{r}
ten_lang <- slice_max(aboriginal_lang, n = 10, order_by = mother_tongue)
ten_lang
```

We will make a bar plot using the `most_at_home` and `language` columns from our `ten_lang` data frame. To create a bar plot of these two variables using the `ggplot` function, we would the following: 

> ```{r img-ggplot, echo = FALSE, message = FALSE, warning = FALSE, fig.cap = "Creating a bar plot with the ggplot function", out.width="1100", fig.retina = 2}

Reviewer: This figure is really awesome and pretty 🤩

Author: Thank you! 👍
